### PR TITLE
Vs eng map shifting lv 408

### PR DESCRIPTION
**Pull request by AlecatEnjoyer**
Jira: https://bradleycapstone.atlassian.net/browse/LV-408?atlOrigin=eyJpIjoiY2Q5ODdjODFiMTk3NDRmMzgxMDUyZTJjZTJkNDhjNDYiLCJwIjoiaiJ9

This push is for making the boat have different layouts as you progress through the game. This should only include the function of it not the criteria for now

### CODE REVIEW

File: Assets/General/Scripts/BoatLayout/MapShifting.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    /// </summary>
+    void Start()
+    {
+        _listOfLayouts[_currentLayout].SetActive(true);
+        Subscribe();
+    }
+    /// <summary>
+    /// Turns off current layout of boat and goes to the next one
+    /// </summary>
+    private protected void LayoutSwapForward()
```
Comment (RyanASwanson) [inline Assets/General/Scripts/BoatLayout/MapShifting.cs:33]: If you have a swap forward you might want to include a swap backwards. Not certain if this is needed as I don't know if we would really ever NEED that functionality but more options are nice. Just a suggestion, not needed

**Review by RyanASwanson — CHANGES_REQUESTED** [review]
Quite clean. Everyone's code keeps getting better and better so I have to be more and more nitpicky in my reviews to compensate. :D

File: Assets/General/Scripts/BoatLayout/MapShifting.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Events;
+
+public class MapShifting : MonoBehaviour
+{
+    //A way of adding layouts into the scene outside of this script
+    [Tooltip("Add all layouts of boat inside of here")]
+    [SerializeField]private protected GameObject[] _listOfLayouts;
```
[the remark this answers is not in the document]
Comment (ASLH68) [reply Assets/General/Scripts/BoatLayout/MapShifting.cs:17]: Maybe rename this to `_boatLayouts`

File: Assets/General/Scripts/BoatLayout/MapShifting.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    /// <summary>
+    /// Turns off current layout of boat and goes all the way back to the
+    /// Original
+    /// </summary>
+    private protected void LayoutSwapToOriginal()
+    {
+        _listOfLayouts[_currentLayout].SetActive(false);
+        _currentLayout = 0;
+        _listOfLayouts[_currentLayout].SetActive(true);
+    }
```
Comment (ASLH68) [inline Assets/General/Scripts/BoatLayout/MapShifting.cs:48]: Add a new line between each function

File: Assets/General/Scripts/BoatLayout/MapShifting.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        _listOfLayouts[_currentLayout].SetActive(true);
+        Subscribe();
+    }
+    /// <summary>
+    /// Turns off current layout of boat and goes to the next one
+    /// </summary>
+    private protected void LayoutSwapForward()
+    {
+        _listOfLayouts[_currentLayout].SetActive(false);
+        ++_currentLayout;
```
Comment (ASLH68) [inline Assets/General/Scripts/BoatLayout/MapShifting.cs:36]: Add a check to make sure that `_currentLayout` is a valid array index

**Review by ASLH68 — CHANGES_REQUESTED** [review]

File: Assets/General/Scripts/BoatLayout/MapShifting.cs
Comment (ASLH68) [inline Assets/General/Scripts/BoatLayout/MapShifting.cs:1]: I think this file should be moved to Assets/GameAssets/World/WorldScripts/Boat/

**Review by ASLH68 — CHANGES_REQUESTED** [review]

File: Assets/General/Scripts/BoatLayout/MapShifting.cs (hunk 1 above)
Comment (AlecatEnjoyer) [reply Assets/General/Scripts/BoatLayout/MapShifting.cs:33]: funny thing is I had one originally but I removed it for some reason

File: Assets/General/Scripts/BoatLayout/MapShifting.cs
Comment (AlecatEnjoyer) [reply Assets/General/Scripts/BoatLayout/MapShifting.cs:1]: I was thinking the same thing when making it

**Review by astapay — APPROVED** [review]
Get the other comments done and we good.

File: Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    {
+        _boatLayouts[_currentLayout].SetActive(true);
+        
+        SubscribeToEvents();
+    }
+    
+    /// <summary>
+    /// Turns off current layout of boat and goes to the next one
+    /// </summary>
+    private protected void LayoutSwapForward()
```
Comment (ASLH68) [inline Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs:40]: None of these functions need to be protected as there is no inheritance going on

File: Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    {
+        if (_currentLayout <= _boatLayouts.Length - 1)
```
Comment (ASLH68) [inline Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs:42]: The way this is setup, if boat layouts is size 5 and current layout is 4, 4 will deactivate, increment to 5 which is an invalid index, then attempt to set 5 to active. This conditional should be immediately before the last line in this function

**Review by ASLH68 — CHANGES_REQUESTED** [review]

**Review by RyanASwanson — APPROVED** [review]
Just address Andreas comments and I approve

File: Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    /// Allow for first layout to turn on if not on already
+    /// </summary>
+    void Start()
+    {
+        _boatLayouts[_currentLayout].SetActive(true);
+        
+        SubscribeToEvents();
+    }
+/// <summary>
+/// THIS IS FOR TESTING ONLY
```
Comment (ASLH68) [inline Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs:38]: Works great! Remove this and the test gym before merging (i know you know this just putting it "on paper" here)

File: Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    
+    //For calling outside of script into another script
+    private UnityEvent _onMapShifting = new();
+    private UnityEvent _onMapReset = new();
+    private UnityEvent _onMapPrevious = new();
+    
+    /// <summary>
+    /// Allow for first layout to turn on if not on already
+    /// </summary>
+    void Start()
```
Comment (ASLH68) [inline Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs:31]: private specifier

File: Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        
+            _boatLayouts[_currentLayout].SetActive(true);
+        }
+    }
+    /// <summary>
+    /// Choose any layout
+    /// </summary>
+    /// <param name="_choice"></param>
+    public void LayoutChoice(int _choice)
+    {
```
Comment (ASLH68) [inline Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs:76]: add a check here for if choice is a valid layout index

File: Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+            ++_currentLayout;
+        
+            _boatLayouts[_currentLayout].SetActive(true);
+        }
+    }
+    /// <summary>
+    /// Choose any layout
+    /// </summary>
+    /// <param name="_choice"></param>
+    public void LayoutChoice(int _choice)
```
Comment (ASLH68) [inline Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs:75]: parameter names do not start with an _

**Review by ASLH68 — CHANGES_REQUESTED** [review]

File: Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        
+            _boatLayouts[_currentLayout].SetActive(true);
+        }
+    }
+    
+    /// <summary>
+    /// Choose any layout
+    /// </summary>
+    /// <param name="_choice"></param>
+    public void LayoutChoice(int choice)
```
Comment (ASLH68) [inline Assets/GameAssets/World/WorldScripts/Boat/MapShifting.cs:76]: suggestions - rename this to `SwapLayout`

**Review by ASLH68 — APPROVED** [review]

**Pull request merged**